Save solution images into artist and playlab level files

davidsbailey opened the pull request:
### Background

For stage extras (see [spec](https://docs.google.com/document/d/1kutbrJTfHAgRoHsp-f0DW-NA_gDemld0s5JC5bBt2Xw/edit)), we want to be able to show images of the bonus levels that the students can choose from. These images should show the solution of the level in question. These levels could be artist or playlab levels. A different solution is already being used to obtain these images for maze levels.

### Description

This PR proposes saving the image at the time that the levelbuilder edits the level solution. This guarantees that the solution image stays up to date when the solution is edited, and puts the solution image url in a predictable place in the database where it can be retrieved when needed. It also has the downside that each of the ~100 levels that will be included in stage extras will need to have their solutions manually run in levelbuilder once in order to generate the solution image and url.

### Verification

I've run one artist solution and one playlab solution to generate the thumbnail image urls for those levels. You can see them here: 

* artist script level: https://studio.code.org/s/coursec/stage/6/puzzle/8
    * level file changes: https://github.com/code-dot-org/code-dot-org/pull/16840/commits/c973dee40f0379f70fcde803e967e47c6564b5f3
    * thumbnail url: https://d3p74s6bwmy6t9.cloudfront.net/8cbbc4899e479126d6c97114489d398f=development/967.png

* playlab script level: https://studio.code.org/s/coursec/stage/13/puzzle/1
    * level file changes: https://github.com/code-dot-org/code-dot-org/pull/16840/commits/9c8934f395cf7ca39c586bbd7bbdd20f7d83222c
    * solution image url: https://d3p74s6bwmy6t9.cloudfront.net/af4338d3357b7926d7966836de647c0d=development/952.png

### Future work

This PR only generates solution images for artist and playlab levels. A slightly different approach may be needed to obtain the actual image for other level types, but the solution image url will still be stored in the same way in the level file for those other level types.

davidsbailey commented:
@aoby , can you please review the rails code? @islemaster , can you please review the javascript? @joshlory , please take a look at the overall approach to make sure that this meets your needs.

islemaster commented on apps/src/turtle/turtle.js:
```diff
@@ -1820,6 +1812,37 @@ Artist.prototype.checkAnswer = function () {
+    return {
+      ...reportData,
+      image,
+    };
```
:+1: for not mutating arguments!

davidsbailey commented:
For clarification, the level file edited in this PR can be seen here in the product: https://studio.code.org/s/coursec/stage/6/puzzle/8 , and the solution image linked from the PR description represents the solution to that level.

ashercodeorg commented on dashboard/test/controllers/levels_controller_test.rb:
```diff
@@ -298,6 +305,36 @@ class LevelsControllerTest < ActionController::TestCase
+    }
+    assert_response :success
+    level = assigns(:level)
+    assert_equal level.properties[:toolbox_blocks.to_s], @program
```
It should be `assert_equal expected, actual`, it appears as if you have the arguments flipped.

aoby replied:
Also why `:toolbox_blocks.to_s`? If you need the string, just specify it directly: `'toolbox_blocks'`. Same below

ashercodeorg commented on dashboard/test/controllers/levels_controller_test.rb:
```diff
@@ -298,6 +305,36 @@ class LevelsControllerTest < ActionController::TestCase
+      game_id: @level.game.id,
+      type: 'toolbox_blocks',
+      program: @program,
+      image: 'stub-image-data',
```
These parameters appear to be (mostly) duplicated across tests. It might make sense to define a `default_params` in the setup block, then modify it appropriately for each test. In addition to being DRY, it makes it more clear what part of the param is under test.

davidsbailey replied:
I've extracted default params for the 5 tests pertaining to update_blocks.

ashercodeorg commented on dashboard/app/helpers/levels_helper.rb:
```diff
@@ -770,4 +770,19 @@ def can_view_teacher_markdown?
     standalone || current_user.try(:should_see_inline_answer?, @script_level)
   end
+
+  def get_level_source_image(level_image, level_source_id)
```
Please document the function. Is `level_image` a string? That `level_source_id` is an integer or nil. It returns a `LevelSourceImage` or nil. Etc.

aoby commented on dashboard/app/helpers/levels_helper.rb:
```diff
@@ -770,4 +770,19 @@ def can_view_teacher_markdown?
     standalone || current_user.try(:should_see_inline_answer?, @script_level)
   end
+
+  def get_level_source_image(level_image, level_source_id)
```
Please add a [function header](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md) comment here. This function is rather complex, and has some side effects I wouldn't necessarily expect from a `get_` method.

I'm not clear to me what it's doing. It seems to take an image, save it to S3, and construct a `LevelSourceImage` model around that? Or if a `LevelSourceImage` already exists, then it finds and returns that and ignores the supplied image?

davidsbailey replied:
renamed to `find_or_create_level_source_image`

aoby commented on dashboard/test/controllers/levels_controller_test.rb:
```diff
@@ -13,6 +13,13 @@ class LevelsControllerTest < ActionController::TestCase
+
+    CDO.stubs(:disable_s3_image_uploads).returns(false)
+    LevelSourceImage.any_instance.expects(:save_to_s3).never
+    create(:level_source, :with_image, level: @level, data: @program)
```
What does the comment relate to? Are we ensuring that we get a real S3 url? I don't see anything to that effect.

davidsbailey replied:
The preceding comment relates to these 3 lines. I was trying to only [tell you why](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/), but maybe I took it too far. I've extracted a method for this and documented the steps more clearly.

davidsbailey commented:
@aoby @ashercodeorg PTAL. I've added comments and addressed other feedback.

aoby commented on dashboard/test/controllers/levels_controller_test.rb:
```diff
@@ -3,6 +3,8 @@
 class LevelsControllerTest < ActionController::TestCase
   include Devise::Test::ControllerHelpers
 
+  default_update_blocks_params = nil
```
This won't be accessible anywhere. Member variables in ruby must be preceded with a `@`, and code here in the class will be executed when the class is loaded, not instantiated.

Since `setup` runs before each test case you don't need to initialize the member, just define it directly. If you want to use it without the `@` you can add an accessor with `attr_reader :default_update_blocks_params`, or instead of in setup define it as a private method so it's lazily constructed as in this [example](https://github.com/code-dot-org/code-dot-org/blob/19d6403d6035f4a3cfa101a5a10bd23e8627304b/dashboard/app/controllers/pd/workshop_enrollment_controller.rb#L242-L241).

davidsbailey replied:
I've followed your suggestion, but FYI that tests were passing for some reason the way that I had it. Not sure I can explain why exactly.

aoby replied:
Weird. I played around with it and confirmed that it worked. But not for the reason it seems. This has to do with the closures generated for the blocks sent to the `test` command when the class is defined. 

See this example:
```ruby
class C
  temp_var = 'value'

  define_method :with_closure do
    puts "with closure: #{defined?(temp_var)}"
  end

  def without_closure
    puts "without closure: #{defined?(temp_var)}"
  end

  define_method :nested do
    print "nested, "
    without_closure
  end
end

c = C.new
c.with_closure
c.without_closure
c.nested
```
The output is:
```
with closure: local-variable
without closure:
nested, without closure:
```

It shows up in the closure as if it were a local variable, but doesn't show up anywhere else.
That is only AFAICT because the blocks were defined in the same context. That variable is only in scope while the class is loaded, not accessible anywhere else later.

Similarly,
```ruby
class C
  @outside = nil

  def initialize
    @inside = nil    
  end

  def m
    puts "outside: #{defined?(@outside)}, inside: #{defined?(@inside)}"
  end
end

C.new.m
```
outputs
```
outside: , inside: instance-variable
```
The instance variable only persists when it's defined in a method. The one defined in the class body is lost.

davidsbailey replied:
cool. TIL

davidsbailey commented:
Quick update: this is now working in playlab also, since we are already sending the feedback image to the server when the solution blocks are updated in playlab [here](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/studio/studio.js#L3387). I've added a solution image url for a single playlab level to this PR, and updated the PR description to point to it.
